ee/codegen/src/generators/workflow.ts

```diff
@@ -636,14 +663,161 @@ export class Workflow {
     });
   }
 
+  private buildGraphAttribute(): AstNode | undefined {
+    if (!this.breadthFirstGraphGeneration) {
+      return this.buildGraphLegacy({
+        portContexts: this.entrypointPortContexts,
+      });
+    }
+
+    // This graph attribute generation is a breadth-first traversal of the WorkflowEdges
+    // starting at the entrypoint port edges. We process each one sequentially and incrementally
+    // build up the graph as we go.
+    type GraphEmpty = { type: "empty" };
+    type GraphSet = { type: "set"; values: GraphMutableAst[] };
+    type GraphNodeReference = {
+      type: "node_reference";
+      reference: BaseNodeContext<WorkflowDataNode>;
+    };
+    type GraphPortReference = {
+      type: "port_reference";
+      reference: PortContext;
+    };
+    type GraphRightShift = {
+      type: "right_shift";
+      lhs: GraphMutableAst;
+      rhs: GraphMutableAst;
+    };
+    type GraphMutableAst =
```

Comment:
fern's ast nodes are immutable, and our algorithm involves an evolving graph attribute